liquibase-core/src/main/java/liquibase/command/core/TagCommandStep.java

```diff
@@ -19,34 +31,53 @@ public class TagCommandStep extends AbstractCliWrapperCommandStep {
-                .description("The JDBC database connection URL").build();
+        TAG_ARG = builder.argument("tag", String.class).required()
+                .description("Tag to add to the database changelog table").build();
+        URL_ARG = builder.argument(CommonArgumentNames.URL, String.class).description("The JDBC database connection URL").build();
         DEFAULT_SCHEMA_NAME_ARG = builder.argument("defaultSchemaName", String.class)
                 .description("The default schema name to use for the database connection").build();
         DEFAULT_CATALOG_NAME_ARG = builder.argument("defaultCatalogName", String.class)
                 .description("The default catalog name to use for the database connection").build();
         DRIVER_ARG = builder.argument("driver", String.class)
                 .description("The JDBC driver class").build();
         DRIVER_PROPERTIES_FILE_ARG = builder.argument("driverPropertiesFile", String.class)
-               .description("The JDBC driver properties file").build();
+                .description("The JDBC driver properties file").build();
         USERNAME_ARG = builder.argument(CommonArgumentNames.USERNAME, String.class)
                 .description("Username to use to connect to the database").build();
         PASSWORD_ARG = builder.argument(CommonArgumentNames.PASSWORD, String.class)
                 .description("Password to use to connect to the database")
                 .setValueObfuscator(ConfigurationValueObfuscator.STANDARD)
                 .build();
-        TAG_ARG = builder.argument("tag", String.class).required()
-                .description("Tag to add to the database changelog table").build();
+        DATABASE_ARG = builder.argument("database", Database.class).hidden().build();
     }
 
+
     @Override
-    public String[][] defineCommandNames() {
-        return new String[][] { COMMAND_NAME };
+    public void run(CommandResultsBuilder resultsBuilder) throws Exception {
+        CommandScope commandScope = resultsBuilder.getCommandScope();
+        this.setOrCreateDatabase(commandScope, DATABASE_ARG);
+        LockService lockService = LockServiceFactory.getInstance().getLockService(getDatabase());
+        lockService.waitForLock();
+
+        try {
+            ChangeLogHistoryServiceFactory.getInstance().getChangeLogService(getDatabase()).generateDeploymentId();
+            checkLiquibaseTables(false, null, new Contexts(), new LabelExpression());
+            getDatabase().tag(commandScope.getArgumentValue(TagCommandStep.TAG_ARG));
+        } finally {
+            try {
+                lockService.releaseLock();
+            } catch (LockException e) {
+                Scope.getCurrentScope().getLog(getClass()).severe(Liquibase.MSG_COULD_NOT_RELEASE_LOCK, e);
+            }
+            if (commandScope.getArgumentValue(DATABASE_ARG) == null) {
+                closeDatabase(false);
```

Comment:
Connection management patterns is a good question for commands. I think this setup is right, where you keep the connection open if the connection is passed, and close it if the connection info is given.